### PR TITLE
[RFC] eval: add api_info()

### PR DESCRIPTION
**Pull request by bfredl**
`v:api_info` instead of `msgpackparse(systemlist('nvim --api-info'))[0]` and also avoids "special" dicts (not that I understand why there where there in this case to begin with, the names are all STR and free of NUL bytes).

### CODE REVIEW

**Comment by bfredl** [comment]
added docs and basic test.

**Comment by justinmk** [comment]
Sure it's not a huge amount, but isn't this keeping a fairly large dict in memory that will almost never be used? Assuming it is only useful for documentation, it would be better to auto-generate a doc section.

> also avoids "special" dicts (not that I understand why there where there in this case to begin with, the names are all STR and free of NUL bytes).

Let's fix that :)

**Comment by mhinz** [comment]
> Sure it's not a huge amount, but isn't this keeping a fairly large dict in memory that will almost never be used? Assuming it is only useful for documentation, it would be better to auto-generate a doc section.

Good point.

**Comment by bfredl** [comment]
Well it was already deemed useful enough for `msgpackparse(systemlist('nvim --api-info'))[0]` to be suggested to the docs. The obvious alternative is to make it a function `f_api_info()` if we are worried about those kilobytes.

File: runtime/doc/msgpack_rpc.txt [hunk 1]
```diff
@@ -59,6 +59,8 @@ There are two ways to obtain API metadata:
    for clients written in statically-typed languages, which require a separate
    compilation step.
 
+3. In vimscript the metadata is available as |v:api_metadata|.
```
Comment (jamessan) [inline runtime/doc/msgpack_rpc.txt]: > as |v:api_info|

**Comment by justinmk** [comment]
`msgpackparse(systemlist('nvim --api-info'))[0]` doesn't seem like a huge burden given the infrequency that this will be needed, plus it exposes the user to a number of techniques and features.

It's not the space that necessarily matters, but don't we usually avoid adding functions and features for things that can be served by trivial plugins? 

And again, an auto-generated doc would definitely be warranted, not to mention confirming whether @ZyX-I intended the behavior we're observing with msgpackparse().

**Comment by bfredl** [comment]
And remember nvim is a perfectly legitimate remote API client to nvim, so this has clearly uses beyond documentation. For documentation we might actually want to extract the full doxygen api docs to `.txt` so one just `:h vim_some_api_function` without having the source code available.

**Comment by bfredl** [comment]
`msgpackparse(systemlist('nvim --api-info'))[0]` will work until nvim is not the first nvim in PATH. Surely it could be made even more complex to always work, but surely there is better ways to introduce techniques and features than actively hiding information because we cannot know how much it will be used in advance.

**Comment by justinmk** [comment]
How about if `v:api` contains actual funcrefs which make API calls? 

    call v:api.vim_some_api_function(...)

> msgpackparse(systemlist('nvim --api-info'))[0] will work until nvim is not the first nvim in PATH. 

`v:progname` and friends could be used, but yes it sounds like we need `f_api_info()`.

**Comment by bfredl** [comment]
Updated to `api_info()`. Three lines of code doesn't seem as much code bloat to worry about :)

**Pull request merged**